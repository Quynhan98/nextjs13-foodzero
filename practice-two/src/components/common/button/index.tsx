import { MouseEvent } from 'react';
import './button.css';

export type ButtonProp = {
  color: 'btn-primary' | 'btn-secondary' | 'btn-warning';
  typeButton: 'button' | 'submit' | 'reset';
  children: string | JSX.Element;
  size: 'btn-small' | 'btn-large';
  handleClick?: (e: MouseEvent<HTMLButtonElement>) => void;
};

export const Button = (props: ButtonProp): JSX.Element => {
  const { color, children, typeButton, size, handleClick } = props;

  return (
    <button type={typeButton} className={`button ${color} ${size}`} onClick={handleClick}>
      {children}
    </button>
  );
};
