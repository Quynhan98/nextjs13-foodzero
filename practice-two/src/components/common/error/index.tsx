import './error.css';

type ErrorProp = {
  children: string | JSX.Element;
  onHandleClick: () => void;
};

export const Error = (props: ErrorProp): JSX.Element => {
  const { children, onHandleClick } = props;

  return (
    <div data-testid="error" className="notification-error">
      <button data-testid="btn-close-err" className="icon-close__error" onClick={onHandleClick}>
        x
      </button>
      {children}
    </div>
  );
};
