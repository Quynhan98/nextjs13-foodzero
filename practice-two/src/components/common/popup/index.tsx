import { MouseEvent } from 'react';
import './popup.css';

export type PopupProp = {
  children: JSX.Element;
  handleClose: (e: MouseEvent) => void;
};

export const Popup = (props: PopupProp): JSX.Element => {
  const { children, handleClose } = props;

  return (
    <div data-testid="popup" className="popup">
      <div className="popup-container">
        <button className="close-icon" onClick={handleClose}>
          &#10008;
        </button>
        {children}
      </div>
    </div>
  );
};
