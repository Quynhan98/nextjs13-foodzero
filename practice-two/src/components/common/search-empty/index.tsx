import searchEmpty from 'assets/images/search-empty.png';
import './search-empty.css';

export const SearchEmptyResult = (): JSX.Element => {
  return (
    <div data-testid="search-empty" className="search-empty">
      <img className="search-empty_icon" src={searchEmpty} alt="search empty result" />
      <p className="search-empty_title">No result is found</p>
      <p className="search-empty_hint">Try using more generic keywords</p>
    </div>
  );
};
