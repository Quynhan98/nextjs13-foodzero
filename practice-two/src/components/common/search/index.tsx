import { ChangeEventHandler, memo } from 'react';
import './search.css';

interface SearchProps {
  onChange?: ChangeEventHandler<HTMLInputElement>;
}

const Search = (props: SearchProps) => {
  const { onChange } = props;

  return (
    <div className="filter-search">
      <input
        data-testid="search"
        type="search"
        placeholder="Search..."
        className="search-input"
        onChange={onChange}
      />
    </div>
  );
};

export default memo(Search);
