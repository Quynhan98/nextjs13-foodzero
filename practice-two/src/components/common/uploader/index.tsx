import { convertBase64 } from '@helpers/encode';
import { FormEvent, useState } from 'react';
import './uploader.css';

export type UploaderProp = {
  label: string;
  messageErr?: string;
  name: string;
  baseImage: string;
  handleChange: (img: string) => void;
  onHandleBlur?: (e: FormEvent) => void;
};

export const Uploader = (props: UploaderProp) => {
  const { label, messageErr, name, baseImage, handleChange, onHandleBlur } = props;
  const [image, setImage] = useState(baseImage);

  const uploadImage = async (e: any) => {
    const file = e.target.files[0];

    const srcImage = (await convertBase64(file)) as string;

    setImage(srcImage);
    handleChange(srcImage);
  };

  return (
    <div className="uploader">
      <label className="uploader-title">
        {label}
        <div className="uploader-input">
          <input
            data-testid="uploader-img"
            onBlur={onHandleBlur}
            name={name}
            onChange={uploadImage}
            className="uploader-content"
            type="file"
          />
          {image && <img data-testid="image-input" src={image} className="uploader-image" />}
        </div>
      </label>
      {messageErr && (
        <span data-testid="img-error" className="input-error">
          {messageErr}
        </span>
      )}
    </div>
  );
};
