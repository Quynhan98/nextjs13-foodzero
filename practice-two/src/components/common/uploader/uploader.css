.uploader-title {
  padding-top: 15px;
  font-weight: bold;
  font-size: var(--size-s);
}

.uploader-content {
  padding: 10px;
  margin-top: 5px;
  font-size: var(--size-s);
  padding-left: 0px;
}

.uploader-input {
  display: flex;
  justify-content: space-between;
}

.uploader {
  padding-top: 15px;
}

.uploader-image {
  width: 100px;
  height: 130px;
}
