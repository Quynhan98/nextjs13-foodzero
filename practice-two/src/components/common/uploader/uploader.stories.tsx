import { Story } from '@storybook/react';
import { Uploader, UploaderProp } from '@components/common/uploader';

export default {
  title: 'Components/Uploader',
  component: Uploader,
};

const Template: Story<UploaderProp> = (args) => <Uploader {...args} />;

export const TextboxInput = Template.bind({});

TextboxInput.args = {
  label: 'Title',
  name: 'image',
};
