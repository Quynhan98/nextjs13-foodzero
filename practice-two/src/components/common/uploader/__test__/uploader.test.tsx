import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Uploader, UploaderProp } from '@components/common/uploader';
import '@testing-library/jest-dom/extend-expect';

describe('Test Uploader Component', () => {
  const handleChangeImage = jest.fn();
  const handleBlur = jest.fn();

  const initialValue: UploaderProp = {
    label: 'Title',
    messageErr: '',
    name: 'book-title',
    baseImage: 'base_img',
    handleChange: handleChangeImage,
    onHandleBlur: handleBlur,
  };

  const initialValueError: UploaderProp = {
    label: 'Title',
    messageErr: 'Image error',
    name: 'book-image',
    baseImage: '',
    handleChange: handleChangeImage,
    onHandleBlur: handleBlur,
  };

  test('Component Uploader matchers DOM Snapshot', () => {
    const component = render(<Uploader {...initialValue} />);

    expect(component).toMatchSnapshot();
  });

  test('Component Uploader with DOM Snapshot match error message', () => {
    const component = render(<Uploader {...initialValueError} />);

    expect(component).toMatchSnapshot();
  });

  test('Component Uploader calls handleChange function when user enter values', async () => {
    const user = userEvent.setup();
    const { getByTestId } = render(<Uploader {...initialValue} />);
    const file = new File(['hello'], 'hello.png', { type: 'image/png' });
    const input = getByTestId('uploader-img') as HTMLInputElement;

    await user.upload(input, file);
    expect(input.files?.item(0)).toStrictEqual(file);
    expect(input.files).toHaveLength(1);
    expect(input.files![0]).toStrictEqual(file);
  });

  test('Component Uploader calls handleBlur function when user enter values', () => {
    const { getByTestId } = render(<Uploader {...initialValue} />);

    fireEvent.blur(getByTestId('uploader-img'));
    expect(initialValue.onHandleBlur).toHaveBeenCalled();
  });

  test('Component Uploader with error message calls handleChange function when user enter values', async () => {
    const user = userEvent.setup();
    const { getByTestId } = render(<Uploader {...initialValueError} />);
    const file = new File(['hello'], 'hello.png', { type: 'image/png' });
    const input = getByTestId('uploader-img') as HTMLInputElement;

    await user.upload(input, file);
    expect(input.files?.item(0)).toStrictEqual(file);
    expect(input.files).toHaveLength(1);
    expect(input.files![0]).toStrictEqual(file);
  });

  test('Component Uploader with error message calling handlerBlur function when user enters value', () => {
    const { getByTestId } = render(<Uploader {...initialValueError} />);

    fireEvent.blur(getByTestId('uploader-img'));
    expect(handleBlur).toHaveBeenCalled();
    expect(screen.getByTestId('img-error')).toHaveTextContent('Image error');
  });
});
