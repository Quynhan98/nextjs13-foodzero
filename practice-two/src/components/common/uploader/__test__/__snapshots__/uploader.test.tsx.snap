// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Uploader Component Component Uploader matchers DOM Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="uploader"
      >
        <label
          class="uploader-title"
        >
          Title
          <div
            class="uploader-input"
          >
            <input
              class="uploader-content"
              data-testid="uploader-img"
              name="book-title"
              type="file"
            />
            <img
              class="uploader-image"
              data-testid="image-input"
              src="base_img"
            />
          </div>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="uploader"
    >
      <label
        class="uploader-title"
      >
        Title
        <div
          class="uploader-input"
        >
          <input
            class="uploader-content"
            data-testid="uploader-img"
            name="book-title"
            type="file"
          />
          <img
            class="uploader-image"
            data-testid="image-input"
            src="base_img"
          />
        </div>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Uploader Component Component Uploader with DOM Snapshot match error message 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="uploader"
      >
        <label
          class="uploader-title"
        >
          Title
          <div
            class="uploader-input"
          >
            <input
              class="uploader-content"
              data-testid="uploader-img"
              name="book-image"
              type="file"
            />
          </div>
        </label>
        <span
          class="input-error"
          data-testid="img-error"
        >
          Image error
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="uploader"
    >
      <label
        class="uploader-title"
      >
        Title
        <div
          class="uploader-input"
        >
          <input
            class="uploader-content"
            data-testid="uploader-img"
            name="book-image"
            type="file"
          />
        </div>
      </label>
      <span
        class="input-error"
        data-testid="img-error"
      >
        Image error
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
